File: NCNN.py
import torch
import torch.nn as nn
import torch.nn.functional as F
from math import floor

class NCONV(nn.Module):
    def __init__(self, kernel_size=1, dilation=1, padding=1, stride=1,
                 filters=None, channel=None, center=True, scale=True, bias=True):
        super(LCONV, self).__init__()

        self.kernel_size = kernel_size
        self.dilation = dilation
        self.padding = padding
        self.stride = stride
        self.filters = filters
        self.channel = channel
        self.center_bool = center
        self.scale_bool = scale

        # Define weights and parameters
        self.weight = nn.Parameter(torch.Tensor(filters, channel, kernel_size, kernel_size))
        self.bias = nn.Parameter(torch.Tensor(filters)) if bias else None
        self.scale = nn.Parameter(torch.Tensor(filters)) if scale else None

        # Initialize weights
        nn.init.kaiming_normal_(self.weight, mode='fan_out')

    def forward(self, x):
        h1, w1 = x.size(2), x.size(3)

        # Apply unfold operation
        x = F.unfold(x, (self.kernel_size, self.kernel_size),
                     dilation=self.dilation, padding=self.padding, stride=self.stride)
        x = x.transpose(1, 2)  # (batch, num_patches, patch_size)

        # Compute mean and variance, then normalize
        if self.center_bool:
            mean = x.mean(dim=-1, keepdim=True)
            x = x - mean
        if self.scale_bool:
            var = x.var(dim=-1, unbiased=False, keepdim=True)
            x = x / torch.sqrt(var + 1e-14)

        # Apply weights
        x = x.matmul(self.weight.view(self.weight.size(0), -1).t())

        # Add bias
        if self.bias is not None:
            x = x + self.bias

        # Reshape back to original dimensions
        x = x.transpose(1, 2)
        h, w = self.conv_output_shape(h1, w1, self.kernel_size, self.stride, self.padding, self.dilation)
        x = x.view(-1, self.filters, h, w)

        return x

    @staticmethod
    def conv_output_shape(h1, w1, kernel_size, stride, pad, dilation):
        """ Compute the output shape after convolution operation """
        h = floor((h1 + 2 * pad - dilation * (kernel_size - 1) - 1) / stride + 1)
        w = floor((w1 + 2 * pad - dilation * (kernel_size - 1) - 1) / stride + 1)
        return h, w

File: NCNN.cpp
#include <torch/extension.h>
#include <iostream>
#include <vector>
std::vector<at::Tensor> ncnn_forward(
    torch::Tensor input,
    torch::Tensor weights,
    torch::Tensor bias,
    torch::Tensor scale,
    torch::Tensor kernel_size,
    torch::Tensor dilation,
    torch::Tensor padding,
    torch::Tensor stride,
    torch::Tensor filters,) {
    
  const auto output_h = (input.size(2) + (2*padding[0]) - ((dilation[0]*kernel_size[0]-1)+1))/stride[0] + 1
  const auto output_w = (input.size(3) + (2*padding[1]) - ((dilation[1]*kernel_size[1]-1)+1))/stride[1] + 1
      
  auto X = torch::unfold(x,kernel_size,dilation=dilation,padding=padding,stride=stride).transpose(1,2);
  
  auto X_mean = torch::mean(X,/*dim=*/-1,/*keepdim=*/true);
      
  auto X_shift = X - X_mean;
  auto X_std = torch::std(X_shift,/*dim=*/-1,/*keepdim=*/true).expand_as(X);
  auto X_refined = X_shift / (X_std+1e-5);
    
  auto X_scaled = scale*X_refined
  
  
  auto flatten = weights.view(weights.size(0),-1).transpose();
  auto output = torch::addmm(bias,X_scaled,flatten).view(-1,filters,output_h,output_w);

    
    
    
  return {output,
          flatten,
          X,
          X_mean,
          X_shift,
          X_std,
          X_refined};
}


std::vector<torch::Tensor> ncnn_backward(
    torch::Tensor grad_output,
    torch::Tensor kernel_size,
    torch::Tensor stride,
    torch::Tensor padding,
    torch::Tensor X,
    torch::Tensor X_mean,
    torch::Tensor X_std,
    torch::Tensor X_refined,
    torch::Tensor X_scaled,
    torch::Tensor weights) {
    
    
  const auto numb = 1.0/kernel_size[0]*kernel_size[1]*X.size(2);
  auto grad_output_ = grad_ouput*scale
  auto grad_sigma = torch::sum(grad_output_ * (X-X_mean),/*axis=*/=-1,/*keepdim=*/=True)*0.5*(X_std)**(-1.5);
  auto grad_mean = torch::sum(grad_output_*(-1./X_std.expand_as(X)),/*axis=*/=-1,/*keepdim=*/=True) + grad_sigma*numb*2.0
      *torch::sum((X-X_mean),/*axis=*/=-1,/*keepdim=*/=True)*-1;
  auto grad_x = grad_output_ * (1/X_std.expand_as(X))+grad_sigma*2.0*torch::sum((X-X_mean),/*axis=*/=-1,/*keepdim=*/=True)+grad_mean*numb;
  
  auto grad_scale = torch::sum(grad_output*X_refined,/*axis=*/=-1)
  auto grad_bias = torch::sum(grad_output,/*axis=*/=-1)
                               

  auto grad_weights = X_scaled.t().mm(grad_output);
    
  return {grad_weights, grad_bias, grad_scale, grad_x, };
}


